test-results.cake

```diff
@@ -0,0 +1,182 @@
+// This file contains classes used to interpret the result XML that is
+// produced by test runs of the GUI.
+
+using System.Xml;
+
+public abstract class ResultSummary
+{
+    public string OverallResult { get; set; }
+    public int Total { get; set; }
+    public int Passed { get; set; }
+    public int Failed { get; set; }
+    public int Warnings { get; set; }
+    public int Inconclusive { get; set; }
+    public int Skipped { get; set; }
+}
+
+public class ExpectedResult : ResultSummary
+{
+	public ExpectedResult(string overallResult)
+	{
+        if (string.IsNullOrEmpty(overallResult))
+            throw new ArgumentNullException(nameof(overallResult));
+
+		OverallResult = overallResult;
+		// Initialize counters to -1, indicating no expected value.
+		// Set properties of those items to be checked.
+		Total = Passed = Failed = Warnings = Inconclusive = Skipped = -1;
+	}
+}
+
+public class ActualResult : ResultSummary
+{
+    public ActualResult(string resultFile)
+    {
+        var doc = new XmlDocument();
+        doc.Load(resultFile);
+
+        Xml = doc.DocumentElement;
+        if (Xml.Name != "test-run")
+            throw new Exception("The test-run element was not found.");
+
+        OverallResult = GetAttribute(Xml, "result");
+        Total = IntAttribute(Xml, "total");
+        Passed = IntAttribute(Xml, "passed");
+        Failed = IntAttribute(Xml, "failed");
+        Warnings = IntAttribute(Xml, "warnings");
+        Inconclusive = IntAttribute(Xml, "inconclusive");
+        Skipped = IntAttribute(Xml, "skipped");
+    }
+
+    public XmlNode Xml { get; }
+
+    private string GetAttribute(XmlNode node, string name)
+    {
+        return node.Attributes[name]?.Value;
+    }
+
+    private int IntAttribute(XmlNode node, string name)
+    {
+        string s = GetAttribute(node, name);
+        return s == null ? 0 : int.Parse(s);
```

Comment:
Should we throw an exception instead, if `s == null`? Just thinking that zero could be a valid value - and these attributes should always be present, I think?

Reply:
You're right in principle. I had a vague memory that we sometimes omitted the attribute in the past, but it may no longer apply. If you want a true check, I would use -1 because I don't want the constructor of the ActualResult to throw. A -1 would cause an error to be generated when the TestReport compared values anyway.

Reply:
I tried -1 here and got an error. I think that's why I used zero in the first place. This time I dug deeper and found that it's an error in the engine. It only affects `test-run` and Project type `test-suilte` elements, which is why it has gone unnoticed. I'll file a separate issue.

I'll leave in the workaround for now.

Reply:
See issue #904